packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts

```diff
@@ -0,0 +1,147 @@
+export class FindDuplicatesQueryFactory {
+  private readonly logger = new Logger(FindDuplicatesQueryFactory.name);
+
+  constructor(
+    private readonly fieldsStringFactory: FieldsStringFactory,
+    private readonly argsAliasFactory: ArgsAliasFactory,
+  ) {}
+
+  async create<Filter extends RecordFilter = RecordFilter>(
+    args: FindDuplicatesResolverArgs<Filter>,
+    options: WorkspaceQueryBuilderOptions,
+    currentRecord?: PGGraphQLResult,
+  ) {
+    const fieldsString = await this.fieldsStringFactory.create(
+      options.info,
+      options.fieldMetadataCollection,
+      options.objectMetadataCollection,
+    );
+
+    const argsData = this.getFindDuplicateBy<Filter>(
+      args,
+      options,
+      currentRecord,
+    );
+
+    const duplicateCondition = this.buildDuplicateCondition(
+      options.objectMetadataItem,
+      argsData,
+      args.id,
+    );
+
+    const filters = stringifyWithoutKeyQuote(duplicateCondition);
+
+    return `
+      query {
+        ${computeObjectTargetTable(options.objectMetadataItem)}Collection${
+          filters ? `(filter: ${filters})` : ''
+        } {
+          ${fieldsString}
+        }
+      }
+    `;
+  }
+
+  getFindDuplicateBy<Filter extends RecordFilter = RecordFilter>(
+    args: FindDuplicatesResolverArgs<Filter>,
+    options: WorkspaceQueryBuilderOptions,
+    currentRecord?: PGGraphQLResult,
+  ) {
+    const entityKey = `${computeObjectTargetTable(
+      options.objectMetadataItem,
+    )}Collection`;
+
+    const currentRecordResult =
+      currentRecord?.[0]?.resolve?.data?.[entityKey].edges?.[0]?.node;
+
+    if (currentRecordResult) {
+      return currentRecordResult as Record<string, unknown>;
+    }
+
+    return this.argsAliasFactory.create(
+      args.data ?? {},
+      options.fieldMetadataCollection,
+    );
+  }
+
+  buildQueryForExistingRecord(
+    id: string,
+    options: WorkspaceQueryBuilderOptions,
+  ) {
+    return `
+      query {
+        ${computeObjectTargetTable(
+          options.objectMetadataItem,
+        )}Collection(filter: { id: { eq: "${id}" }}){
+          edges {
+            node {
+              ${this.getApplicableDuplicateCriterias(options.objectMetadataItem)
+                .flatMap((dc) => dc.fieldNames)
+                .join('\n')}
+            }
+          }
+        }
+      }
+    `;
+  }
+
+  private buildDuplicateCondition(
+    objectMetadataItem: ObjectMetadataInterface,
+    argsData?: Record<string, unknown>,
+    filteringByExistingRecordId?: string,
+  ) {
+    if (!argsData) {
+      return;
+    }
+
+    const criterias = this.getApplicableDuplicateCriterias(objectMetadataItem);
+
+    return {
+      // when filtering by an existing record, we need to filter that explicit record out
+      ...(filteringByExistingRecordId && {
+        id: { neq: filteringByExistingRecordId },
+      }),
+      // keep condition as "or" to get results by more duplicate criterias
+      or: criterias
+        .map((dc) =>
```

Comment:
Ive tried to improve the readability a bit, but it is harder than it seems 😄 Im open to further suggestions. The other part of this comment regarding an empty array instead of all the results - damn, how do i create a query that resolves zero records without any hacks? 🤔 😆 Cant figure it out.